### ISSUE TITLE
Re-add automatic daily testing

### ISSUE THREAD

**Issue by hellais** [ooni/probe-android#245]
We should re-add the feature for doing automatic daily testing.

The MVP is that of doing automatic testing on a daily basis which is enable/disable.

In the future we may want to provide more configuration options to the user, namely:
- Enabling only on wifi
- Setting data usage caps
- Configuring how many times it runs or time of day

cc @lorenzoPrimi @bassosimone

**Comment by hellais** [comment]
Notes from the meeting:

To this end (as “quick”, interim solutions):

Re-introduce automatic daily testing on OONI Probe Android app => Include in 2.2, 2.3 or 2.4. Release => MVP is to introduce the previous code, and then improve upon the code based on the features that we want. => Limit the testing based on bandwidth (i.e. how much bandwidth the user is willing to donate for testing). => We should provide choice to users: If they’re on Wifi, they can have everything tested automatically (but set a limit there). If they’re on 3G, they can tell us how much bandwidth they’re willing to donate to testing.
Re-introduce reminder for daily testing in OONI Probe iOS app.

In the long-term:

Evaluate whether to add support to apps for automatic daily testing, or create separate Android app for automatic daily testing (OONI Daily)
Add support to desktop apps for automatic daily testing (scheduling)
Evaluate whether it makes sense to create an Android TV app of sorts.

**Comment by lorenzoPrimi** [comment]
These are the strings we already have (translated) for the upcoming Orchestrate features
<img width="655" alt="Screenshot 2019-09-02 at 21 51 39" src="https://user-images.githubusercontent.com/1616097/64171103-1d8e0e80-ce52-11e9-815a-d1700293dd88.png">
The plan is to have the "Automatic Testing section" and inside that all the above strings.

Now my question is, where do we insert this automatic daily (local) testing option?

Under the automatic testing section of the settings? 
How do we differentiate this feature from the upcoming orchestrate feature?
Which strings should we use?
These are the old 1.x strings for reference:
```
"local_notifications" = "Run test reminder";
"local_notifications_time" = "Time";
"local_notifications_text" = "Time to run tests";

<string name="local_notifications">Run test automatically</string>
<string name="local_notifications_time">Time</string>
 <string name="local_notifications_text">Time to run tests</string>
```

**Comment by hellais** [comment]
> Now my question is, where do we insert this automatic daily (local) testing option?

I think we should expose it inside of a dedicated settings section called "Automated testing" (or maybe @agrabeli has thoughts on a better name). As our automated testing capabilities improve over time we can enrich this section with more options and settings such as those related to push & run, silent push & run, etc.

> How do we differentiate this feature from the upcoming orchestrate feature?

I would not worry about this for the moment, since this feature does not yet exist and we have decided (following the discussions at the meeting in stockholm) that we are going to develop these features incrementally:

1. Add local scheduling for automatic testing as it was in 1.x
2. Push & Run, we send a push notification and the user needs to tap to run an experiment
3. Silent push, we send a push notification and the test is initiated without user input

On top of these 3 we will likely be incrementally also adding more configuration settings such as handling the bandwidth quota.

**Comment by lorenzoPrimi** [comment]
Ok the "Automated testing" section was already planned, we already have strings and icons.
So for the point 1, that is what is being discussed in this issue, which strings should I add?
For iOS I guess "Run test reminder" and the sub-option "Time" is ok, bur what about for Android?
Should we call it "Run test automatically" even if it's a string we added for orchestra?

Furthermore, should we only add this option like it was in 1.x without possibility to configure anything except the time, for now?

**Comment by hellais** [comment]
Here is a quick mock of how these pages can look like.

![Screenshot 2019-09-03 at 15 00 37](https://user-images.githubusercontent.com/424620/64175458-9e9dd380-ce5b-11e9-9013-3df4ff074bdc.png)

Ideally we would also include the ability to pick which tests will be automatically run and which website categories are enabled.

The options below "Run tests automatically" only appear if the toggle is enabled.

"Run tests automatically" should say "Run tests reminder" on iOS.

I will leave the final decision WRT copy to @agrabeli.

**Comment by lorenzoPrimi** [comment]
The categories are already selected in the web connectivity test. 
Should this be a (redundant) link to that section?
"Tests to run" is something new, do we want to add it now or just leave it to run Websites & Middleboxes & IM like the old version?

**Comment by hellais** [comment]
> The categories are already selected in the web connectivity test.
> Should this be a (redundant) link to that section?

I think we should have two different settings for the category selection. A user may want to trigger a test with a different set of categories as the automated testing.

> "Tests to run" is something new, do we want to add it now or just leave it to run Websites & Middleboxes & IM like the old version?

How much effort is it to do this?

**Comment by lorenzoPrimi** [comment]
> How much effort is it to do this?

Not much, it was more a question if we want it now or we wanted the really basic feature. Will add to the settings. 

Now the category selection is unique in the entire app, to duplicate it not only wouldn't be so easy, but also we need to consider it from an UX point of view as the user maybe disable some categories in Web Connectivity then the normal reasoning would be to expect them to be disabled in the automatic Web Connectivity.

**Comment by agrabeli** [comment]
@hellais I agree with @lorenzoPrimi that it probably makes sense to only include the Website category selection only in 1 place, to avoid duplication and confusion. 

That said, perhaps it would be good to move the Websites categories selection option to the main Settings page, as it's quite buried in the settings of the Websites card. 

Under "Automated Testing", I would propose adding a "Test every X hours" option (in addition to the option where you can specify the time of testing, such as 10am). It may be easier for users to say that they're interested in automatic testing once or twice a day, than specifying the precise time of testing (though that may be a useful option for other users).

**Comment by lorenzoPrimi** [comment]
> That said, perhaps it would be good to **move the Websites categories selection option to the main Settings page**, as it's quite buried in the settings of the Websites card.

to move the "Categories" page would be better to open a new issue and discuss it there.

> **Under "Automated Testing", I would propose adding a "Test every X hours" option** (in addition to the option where you can specify the time of testing, such as 10am). It may be easier for users to say that they're interested in automatic testing once or twice a day, than specifying the precise time of testing (though that may be a useful option for other users).

I like the "Test every X hour" but it collides a bit with the "Test it at 6pm" option, at the end this last option isn't just a "Test every 24 hour"?
Maybe there is a way to put these two options together?

> Users would also probably want to **specify how many URLs they test within the scope of automatic testing**. So under the "automated testing" section, we may want to include options like "Use XXX data/bandwidth for testing", "Test full test lists", etc.

The old automatic testing feature takes the options from the current app, so uses the default max runtime. I would leave it like this and open a new issue deciding how to improve it. 
The scope of this issue is just to reimplement what there was in 1.x
Data/bandwidth is also an very advanced option, we might test it here but ideally will come with orchestra

**Comment by hellais** [comment]
> to move the "Categories" page would be better to open a new issue and discuss it there.

Agreed. I filed: https://github.com/ooni/probe/issues/881

> Maybe there is a way to put these two options together?

I agree with @agrabeli that it's maybe ideal to request the user the number of times they wish to run the test per day and only if they require further customisation then we allow them to choose the times.

I would say that in this first iteration we can maybe just focus on having the user choose how many times per day they would like to run the testing. By having them pick the number of times in a day we can also implement in the app side randomisation of testing time, which is important to ensure we have good testing coverage throughout the day and avoid all probes hitting us at the same time (thundering herd problem).

 > The old automatic testing feature takes the options from the current app, so uses the default max runtime. I would leave it like this and open a new issue deciding how to improve it.
> The scope of this issue is just to reimplement what there was in 1.x
> Data/bandwidth is also an very advanced option, we might test it here but ideally will come with orchestra

I agree that adding this feature to this first iteration is perhaps too much and we should defer it to future work, though let's avoid falling into the trap we were falling into in the past of saying "ah no worries orchestration will solve this", because following the most recent discussions we agreed to move forward with orchestration in a more minimalist sense and actually try to achieve as much as we can without relying on a fully blown orchestration.

**Comment by lorenzoPrimi** [comment]
> I would say that in this first iteration we can maybe just focus on having the user choose how many times per day they would like to run the testing. By having them pick the number of times in a day we can also implement in the app side randomisation of testing time, which is important to ensure we have good testing coverage throughout the day and avoid all probes hitting us at the same time (thundering herd problem).

Yeah I never totally liked the run test at 18:00 option as maybe it's not 100% sure that at that time I'll be at home with wifi.
Should we add an option "Run test under Wi-fi" only or schedule it for 2.3.1 ?

> I agree that adding this feature to this first iteration is perhaps too much and we should defer it to future work, though let's avoid falling into the trap we were falling into in the past of saying "ah no worries orchestration will solve this", because following the most recent discussions we agreed to move forward with orchestration in a more minimalist sense and actually try to achieve as much as we can without relying on a fully blown orchestration.

Same to the consideration of above. I'm trying to come up with a plan for 2.3.0, 2.3.1 and so on to ship this feature in small pieces, rather than a big piece of code, so it's also easier to test and implement.

**Comment by hellais** [comment]
> Should we add an option "Run test under Wi-fi" only or schedule it for 2.3.1 ?

I think having this option would be ideal, though I will leave it to you if think it's reasonable to include in 2.3.0 or we should defer to 2.3.1.

Ideally if we do include this option it would be cool to also support specifying the amount of traffic allocated to automated testing.

**Comment by lorenzoPrimi** [comment]
2.3.0
Android:
- Run test automatically
Options:
- How many times per day. Randomize the hours in the day.
- Tests to run
- Only test if wifi connected

iOS
- Run test reminder with choose time

2.3.X
New options for Android:
- Choose different categories (after we solve https://github.com/ooni/probe/issues/881)
- Choose different max_runtime for auto tests
- How many URLs they test (https://github.com/ooni/probe-android/issues/255)
- Data/bandwidth Limits (if this data is exceeded will not run tests)

**Comment by lorenzoPrimi** [comment]
Branches
https://github.com/ooni/probe-android/tree/automated_testing
https://github.com/ooni/probe-ios/tree/automatic_testing
